### PR TITLE
[POC] Add data migrations

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR adds data migrations handled by the gem [rails-data-migrations](https://github.com/OffgridElectric/rails-data-migrations). As I mentioned in https://github.com/thepracticaldev/dev.to/issues/6332 this gem is pretty small, maintained and it hooks into the existing migration logic Rails offers.

This would give us for free a few things:

- versioning builtin (the gem uses the same mechanism that Rails schema migration use)
- the concept of pending migrations
- a few tasks:

```shell
rake data:migrate                             # Apply pending data migrations
rake data:migrate:down                        # Revert single data migration using VERSION
rake data:migrate:pending                     # List pending migrations
rake data:migrate:skip                        # Skip single data migration using VERSION
rake data:migrate:up                          # Apply single data migration using VERSION
rake data:reset                               # Mark all pending data migrations complete
```

(by default data migrations can only go `up`, that's also how they are generated)

I also added a Rails initializer (`development.rb` was starting to get too long) that checks the following two things:

- pending schema migrations
- pending data migrations

It raises exceptions if we have either

I also added a POC (not tested yet) worker that does the same thing `rails data:migrate` does, but asynchronously

- do we need statuses? Right now we only run migrations async in production, and we're not going to retry failed data migrations in productions anyway, so notifying honeybadger might actually be enough

I haven't removed the existing code yet, I would do it in a separate PR anyway.

The migrations files you see in the PR are the existing data update scripts moved over

To test this you can just checkout the PR and run `rails console`, it should tell you that you have to run data migrations ;)

Let me know if there's something huge I'm missing!

TODO:

- [ ] add tests
- [ ] update documentation

## Related Tickets & Documents

- https://github.com/thepracticaldev/dev.to/issues/6332

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ ] no documentation needed

### CODE REVIEW

**Comment by lightalloy** [comment]
I like this approach better than the current one. It makes more sense for me to use an existing solution instead of trying to reimplement migrations in our codebase, taking existing issues described in the issue into an account.
The provided solution seems good enough, I also don't see a strong need for data migrations statuses.

File: app/errors/pending_data_migration_error.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Rails 6 has ActiveSupport::ActionableError which would allow us to tell the
```
Comment (citizen428) [inline app/errors/pending_data_migration_error.rb]: 👍 , I think I also suggested this in one of our issues, would be great to get to Rails 6 😃

File: app/workers/data_migration_worker.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class DataMigrationWorker
+  include Sidekiq::Worker
+  sidekiq_options queue: :high_priority, retry: 5
+
+  def perform
+    RailsDataMigrations::Migrator.list_pending_migrations.sort_by(&:version).each do |m|
```
Comment (citizen428) [inline app/workers/data_migration_worker.rb]: For reference for other reviewers:

https://github.com/OffgridElectric/rails-data-migrations/blob/4b893f9d2e5dd55e198f8edb9076697f004381ae/lib/tasks/data_migrations.rake#L4-L11

**Comment by citizen428** [comment]
@rhymes What's the status on this? Should we close it out for now or will you be working on this again?

**Pull request closed without merging**